psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1785 @@
+    } else {
+        charges_z = C_left_ao_[0]->ncol();
+    }
+    //
+    // D_ao_ and D_
+    current_costs += 3 * nbf_ * nbf_;
+    // Coulomb Metrics
+    current_costs += 2 * naux_ * naux_;
+    current_costs += 4 * nbf_ * nbf_;
+    // adding in costs for U and A
+    current_costs += biggest_shell * naux_ * charges_z;
+    current_costs += biggest_shell * naux_ * charges_z;
+    current_costs += omp_nthread_ * 5 * nbf_;
+
+    if (memory_ - current_costs > memory_) {
+        std::stringstream error_message;
+        error_message << "DirectDFJK: not enough memory for overhead! Overhead requires "
+                      << current_costs * 8 / (1024 * 1024) << " [MiB].";
+        throw PSIEXCEPTION(error_message.str().c_str());
+    }
+
+    Shell_starts_.push_back(0);
+    for (size_t shell_iter = 0; shell_iter < p_shells_; shell_iter++) {
+        shell_funcs = primary_->shell(shell_iter).nfunction();
+        // The 5 is where we're getting our fudge factor we're getting charges_z
+        //   to stand in for n_occ, and there should only be a factor of 2 included for
+        //   the number of tensors with dimension occ. The extra 3 is a massive amount of
+        //   memory that should cover the overhead from the reast of the calculation.
+        //   charges_z also includes a fudge factor
+        if ((block_size + shell_funcs + 2) * naux_ * (2 * charges_z) > memory_ - current_costs) {
+            Block_funcs_.push_back(block_size);
+            block_size = shell_funcs;
+            Shell_stops_.push_back(shell_iter - 1);
+            Shell_starts_.push_back(shell_iter);
+        } else {
+            block_size += shell_funcs;
+            if (block_size > bmf) {
+                bmf = block_size;
+            }
+        }
+    }
+    Shell_stops_.push_back(p_shells_ - 1);
+
+    Block_funcs_.push_back(block_size);
+
+    biggest_block_ = bmf * naux_ * nbf_;
+    num_blocks_ = Block_funcs_.size();
+
+    k_disps_.resize(num_blocks_);
+    size_t col_disp = 0U;
+    size_t row_disp = 0U;
+
+    for (size_t row_it = 0; row_it < num_blocks_; row_it++) {
+        k_disps_[row_it].resize(num_blocks_);
+        for (size_t col_it = 0; col_it < num_blocks_; col_it++) {
+            k_disps_[row_it][col_it] = nbf_ * row_disp + col_disp;
+            col_disp += Block_funcs_[col_it];
+        }
+        row_disp += Block_funcs_[row_it];
+        col_disp = 0U;
+    }
+}
+
+// decides where all the blocks will stop and start
+// initializes Shell_starts_ and Shell_stops_
+// initializes num_blocks_
+void DirectDFJK::prepare_Q_blocks() {
+    //	std::vector<size_t> starts;//Shell_starts_.get();
+    //	std::vector<size_t> stops;//stops_.get();
+    size_t current_costs = 0;
+    double charges_f = 0.0;
+    size_t charges_z = 0;
+
+    std::shared_ptr<Molecule> mol_ptr = primary_->molecule();
+
+    for (int atom_iter = 0; atom_iter < mol_ptr->natom(); atom_iter++) charges_f += mol_ptr->fZ(atom_iter);
+
+    charges_f = charges_f / 2.0;
```

Comment:
repeat of the p-block section? Maybe code repetition could be avoided by moving both parts to the class init?

Reply:
no, this code is distinct for the q-blocking case. It is called in the memory estimator which is itself called in class init iirc.

Reply:
The two bits of code are distinct, and they get called in the memory estimate which iirc is called when the object is initiated.